action.yml

```diff
@@ -2,19 +2,63 @@ name: Install Swift
 description: Install Swift Release
 
 inputs:
+  # For official swift.org builds: specify only branch and tag
+  # For a custom toolchain build from a Github release: specify github-repo ("owner/repo"), branch, tag and github-release-asset-name
   branch:
-    description: 'Swift Build Branch'
+    description: 'Swift build branch for swift.org builds, or Git branch for Github releases'
     required: true
   tag:
-    description: 'Swift Build Tag'
+    description: 'Swift build tag for swift.org builds, or tag ("tags/TAG") or release name for Github releases'
     required: true
+  github-token:
+    description: 'Github token for fetching a release. Only specify when using custom toolchains from Github releases.'
+    required: false
+  github-repo:
+    description: 'Github repo in "owner/repo" format. Only specify when using custom toolchains from Github releases.'
+    required: false
+  github-release-asset-name:
+    description: 'Asset name for the Swift installer executable in the release. Only specify when using custom toolchains from Github releases.'
+    required: false
 
 runs:
   using: 'composite'
 
   steps:
+    - name: Validate inputs
+      id: validation
+      shell: bash
+      run: |
+        if [[ -n "${{ inputs.github-repo }}" && -n "${{ inputs.github-release-asset-name }}" ]]; then
+          if [[ "${{ runner.os }}" == "Linux" && "${{ inputs.github-release-asset-name }}" != *.tar.gz ]]; then
+            echo >&2 "::error::inputs Invalid action inputs: github-release-asset-name has to be a '*.tar.gz' file on Linux platforms."
+            exit 1
+          elif [[ "${{ runner.os }}" == "macOS" && "${{ inputs.github-release-asset-name }}" != *.pkg ]]; then
+            echo >&2 "::error::inputs Invalid action inputs: github-release-asset-name has to be a '*.pkg' file on MacOS platforms."
+            exit 1
+          elif [[ "${{ runner.os }}" == "Windows" && "${{ inputs.github-release-asset-name }}" != *.exe ]]; then
+            echo >&2 "::error::inputs Invalid action inputs: github-release-asset-name has to be a '*.exe' file on Windows platforms."
+            exit 1
+          fi
+          echo "use_custom_url=1" >> $GITHUB_OUTPUT
+        elif [[ -z "${{ inputs.github-repo }}" && -z "${{ inputs.github-release-asset-name }}" ]]; then
+          echo "use_custom_url=0" >> $GITHUB_OUTPUT
+        else
+          echo >&2 "::error::inputs Invalid action inputs: for fetching a custom Swift toolchain, please specify github-repo, branch, tag and github-release-asset-name inputs."
+          echo >&2 "::error::inputs For using the Swift official toolchain, please specify only branch and tag inputs."
+          exit 1
+        fi
+
+    - name: Fetch installer from Github release
+      uses: dsaltares/fetch-gh-release-asset@a40c8b4a0471f9ab81bdf73a010f74cc51476ad4  # pinned commit for v1.1
```

Comment:
Does the repository not have a tag?

Reply:
it does (v1.1), but tags can move. I like to specify the commit SHA as a way to protect from supply chain compromises on actions from third parties - particularly because this action can have access to private github repos. Let me know if you don't like this pattern and if there are alternatives.

Reply:
True.  I suppose that I would defer to what GH recommends.  The commitish is annoying because you cannot easily determine what version it corresponds to and then find the changes in a release.

Reply:
That's why I added the version as a comment.

Reply:
For context: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions